pkg/cluster/defaults.go

```diff
@@ -28,3 +31,49 @@ func (d ControlPlaneIPCheckAnnotationDefaulter) ControlPlaneIPCheckDefault(ctx c
 
 	return spec, nil
 }
+
+// MachineHealthCheckDefaulter is the defaulter created to configure the machine health check timeouts.
+type MachineHealthCheckDefaulter struct {
+	nodeStartupTimeout      string
+	unhealthyMachineTimeout string
+}
+
+// NewMachineHealthCheckDefaulter allows to create a new MachineHealthCheckDefaulter.
+func NewMachineHealthCheckDefaulter(nodeStartupTimeout, unhealthyMachineTimeout string) MachineHealthCheckDefaulter {
+	return MachineHealthCheckDefaulter{
+		nodeStartupTimeout:      nodeStartupTimeout,
+		unhealthyMachineTimeout: unhealthyMachineTimeout,
+	}
+}
+
+// MachineHealthCheckDefault sets the defaults for machine health check timeouts.
+func (d MachineHealthCheckDefaulter) MachineHealthCheckDefault(ctx context.Context, spec *Spec) (*Spec, error) {
+	if spec.Config.Cluster.Spec.MachineHealthCheck != nil && len(spec.Config.Cluster.Spec.MachineHealthCheck.NodeStartupTimeout) != 0 && len(spec.Config.Cluster.Spec.MachineHealthCheck.UnhealthyMachineTimeout) != 0 {
+		return spec, nil
+	}
+
+	if spec.Config.Cluster.Spec.MachineHealthCheck == nil {
+		spec.Config.Cluster.Spec.MachineHealthCheck = &anywherev1.MachineHealthCheck{}
+	}
+
+	if len(spec.Cluster.Spec.MachineHealthCheck.NodeStartupTimeout) == 0 {
+		if spec.Cluster.Spec.DatacenterRef.Kind == anywherev1.TinkerbellDatacenterKind && d.nodeStartupTimeout == constants.DefaultNodeStartupTimeout.String() {
+			d.nodeStartupTimeout = constants.DefaultTinkerbellNodeStartupTimeout.String()
+		}
+	}
+
+	setMachineHealthCheckTimeoutDefaults(spec.Cluster, d.nodeStartupTimeout, d.unhealthyMachineTimeout)
+
+	return spec, nil
+}
+
+// setMachineHealthCheckTimeoutDefaults sets default tiemout values for cluster's machine health checks.
+func setMachineHealthCheckTimeoutDefaults(cluster *anywherev1.Cluster, nodeStartupTimeout, unhealthyMachineTimeout string) {
```

Comment:
ok so if I'm getting this right, this defaulter will always be initialized with something, either user provided values or our constant defaults. Which means that any cluster being managed from the CLI will end up with values set in these fields, no matter if the user set them or not.

This seems to present a change vs the API managed clusters. It seems like those by default will have empty values and then the controller will use the defaults when creating the MHC, but won't alter the original cluster.

I'm not pushing one way or the other, but I would love if we could make the behavior consistent.

Reply:
This also makes the same changes for clusters created with the API. the original cluster object looks like same with MHC initialized in the spec for API and CLI managed clusters.